rfc/plugins.md

```diff
@@ -0,0 +1,300 @@
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+When a service starts up, PF4J will scan a designated locally-accessible directory for plugins to install (these plugins may be downloaded at runtime or baked into the image at built-time). Once an application has started, it will notify Front50 of the plugins that are installed, and get a response dictating which plugins are to be enabled.
+
+In the case of a runtime-downloaded plugin, version to download will be dictated by Front50. Should Front50 be unavailable, the service will download the version defined by the service, or if none is defined, the latest plugin version that matches the Spinnaker service version.
+
+When a new plugin version is released, it is the responsibility of the operator to release this new version. _**TBD**: How will an operator know a new version exists?_ This new version will need to be released in lock-step with the services that it is associated with. In the future, this functionality _may_ be dynamically sourced and then re-activated. See [pf4j/pf4j-update](https://github.com/pf4j/pf4j-update)
+
+Tightly-coupled service and plugin releases is a poor developer and operator story, but an acceptable shortcoming for an initial release. By using Front50 as the source of truth, we’ll be fairly well equipped to make an acceptable runtime plugin loading story.
+
+Assuming a world where in-process plugins can be upgraded at runtime, services will need a background process to download these new plugins whenever a delta is detected in Front50, and the plugin refreshed. PF4J offers an abstraction for this reloading mechanism, but we would need to figure out how Spring integrations are handled. For example, it’s possible we may need to make the constraint that exported beans from a plugin are not refreshed on plugin up/down grade at least initially.
+
+#### Remote
+Remote plugin upgrade process is not much different than iterating a remote service, since that’s what it is.
+
+When a remote plugin is released, its internal functionality will change immediately, just as it would be expected for a remote service. When the plugin starts, it must send a registration event to Spinnaker. This registration event would contain, at least:
+
+* Plugin ID
+* Other plugin metadata (author, description, version, etc).
+* What extension points it integrates with
+
+Upon receiving this registration event, Spinnaker would then update the Front50 registry with the new information and downstream services would re-bootstrap the plugin, adding or removing it from extension points as necessary.
+
+### Deployment of Remote Plugins
+
+Halyard based Spinnaker installations will use halyard to deploy remote plugins, similar to other spinnaker services.
```

Comment:
Oooh, that sounds pretty ambitious. One of the benefits of remote plugins is that they can be coded in whatever stack the developer prefers. I'm unfamiliar with the magic powers of halyard, but I'd guess it's tailored to deploying JVM-based services. How do we plan to account for other types of runtimes?

Reply:
I've removed this statement - there's no technical benefit to hammering remote plugins into a bespoke Spinnaker-only config tool. Not sure how this made it into the proposal. 😬

Reply:
Ah maybe the statement was too broad, I should have specified this is for deploying container based remote plugins in a Kubernetes environment. The statement here was based on that, as Halyard deploys Spinnaker services as containers in Kubernetes environments. 

I mention it in another comment, but I view plugins as part of a Spinnaker runtime environment and Halyard should ensure the runtime environment is as configured. I do think Halyard should deploy container based plugins and should validate non-container based remote plugins are reachable/ready for traffic (eg similar to cloud provider account validation). 

I expect most remote plugin providers will distribute their plugins as containers, as is so very popular in the devops space these days, but time will tell. Do you guys have a similar expectation? Curious about your guys' thoughts on all this.